File: ios/TNMMultipartUploadProxy.swift
//
//  TNMMultipartUploadProxy.swift
//  TiNetworkManager
//
//  Created by Douglas Alves on 02/01/26.
//


/**
 * Ti.Network.Manager - Multipart Upload Proxy
 * Proxy for multipart/form-data uploads with per-file progress
 * Exposed to JavaScript for creating multipart uploads
 */

import TitaniumKit
import Foundation

@objc(TiHTTPMultipartUploadProxy)
class TNMMultipartUploadProxy: TiProxy {
    
    // MARK: - Properties
    
    private var multipartManager: TNMMultipartUploadManager
    private var certificatePinningManager: TNMCertificatePinningManager
    
    private var uploadId: String
    private var url: String
    private var fields: [MultipartField]
    private var headers: [String: String]
    private var priority: String
    
    private var isActive = false
    
    // MARK: - Initialization
    
    init(
        params: [String: Any],
        multipartManager: TNMMultipartUploadManager,
        certificatePinningManager: TNMCertificatePinningManager
    ) {
        self.multipartManager = multipartManager
        self.certificatePinningManager = certificatePinningManager
        
        self.uploadId = UUID().uuidString
        self.url = params["url"] as? String ?? ""
        self.headers = params["headers"] as? [String: String] ?? [:]
        self.priority = params["priority"] as? String ?? "normal"
        
        // Parse fields
        var parsedFields: [MultipartField] = []
        
        if let fieldsArray = params["fields"] as? [[String: Any]] {
            for fieldDict in fieldsArray {
                if let type = fieldDict["type"] as? String {
                    if type == "text" {
                        // Text field
                        if let name = fieldDict["name"] as? String,
                           let value = fieldDict["value"] as? String {
                            parsedFields.append(.text(name: name, value: value))
                        }
                    } else if type == "file" {
                        // File field
                        if let name = fieldDict["name"] as? String,
                           let filename = fieldDict["filename"] as? String,
                           let mimeType = fieldDict["mimeType"] as? String {
                            
                            // Get file data (can be base64 or file path)
                            if let base64 = fieldDict["data"] as? String,
                               let data = Data(base64Encoded: base64) {
                                parsedFields.append(.file(name: name, filename: filename, data: data, mimeType: mimeType))
                            } else if let filePath = fieldDict["file"] as? String,
                                      let fileURL = URL(string: filePath),
                                      let data = try? Data(contentsOf: fileURL) {
                                parsedFields.append(.file(name: name, filename: filename, data: data, mimeType: mimeType))
                            }
                        }
                    }
                }
            }
        }
        
        self.fields = parsedFields
        
        super.init()
        
        TNMLogger.debug("Multipart upload proxy created", feature: "MultipartUpload", details: [
            "uploadId": uploadId,
            "url": url,
            "fieldCount": parsedFields.count,
            "priority": priority
        ])
    }
    
    // MARK: - Public API
    
    /**
     * Start the multipart upload
     * 
     * @param arguments Unused arguments array (required by Titanium)
     */
    @objc(upload:)
    func upload(arguments: [Any]?) {
        guard !isActive else {
            TNMLogger.warning("Multipart upload already active", feature: "MultipartUpload", details: [
                "uploadId": uploadId
            ])
            return
        }
        
        guard let url = URL(string: self.url) else {
            TNMLogger.error("Invalid URL for multipart upload", feature: "MultipartUpload", details: [
                "uploadId": uploadId,
                "url": self.url
            ])
            fireEvent("error", with: ["error": "Invalid URL"])
            return
        }
        
        guard !fields.isEmpty else {
            TNMLogger.error("No fields provided for multipart upload", feature: "MultipartUpload", details: [
                "uploadId": uploadId
            ])
            fireEvent("error", with: ["error": "No fields provided"])
            return
        }
        
        isActive = true
        
        // Convert priority
        let urlPriority: Float
        switch priority {
        case "high":
            urlPriority = URLSessionTask.highPriority
        case "low":
            urlPriority = URLSessionTask.lowPriority
        default:
            urlPriority = URLSessionTask.defaultPriority
        }
        
        // Get certificate validator
        let certificateValidator = certificatePinningManager.getValidator(for: url.host ?? "")
        
        // Start upload
        multipartManager.startUpload(
            uploadId: uploadId,
            url: url,
            fields: fields,
            headers: headers,
            priority: urlPriority,
            certificateValidator: certificateValidator,
            onProgress: { [weak self] progress in
                self?.handleProgress(progress)
            },
            onFileProgress: { [weak self] filename, sent, total in
                self?.handleFileProgress(filename: filename, sent: sent, total: total)
            },
            onComplete: { [weak self] statusCode, headers, data in
                self?.handleComplete(statusCode: statusCode, headers: headers, data: data)
            },
            onError: { [weak self] error in
                self?.handleError(error)
            }
        )
    }
    
    /**
     * Cancel the upload
     * 
     * @param arguments Unused arguments array (required by Titanium)
     */
    @objc(cancel:)
    func cancel(arguments: [Any]?) {
        guard isActive else { return }
        
        TNMLogger.info("Cancelling multipart upload", feature: "MultipartUpload", details: [
            "uploadId": uploadId
        ])
        
        multipartManager.cancelUpload(uploadId: uploadId)
        isActive = false
        
        fireEvent("cancelled", with: [:])
    }
    
    // MARK: - Event Handlers
    
    private func handleProgress(_ progress: MultipartUploadProgress) {
        guard isActive else { return }
        
        var eventData: [String: Any] = [
            "sent": progress.totalBytesSent,
            "total": progress.totalBytesExpected,
            "progress": progress.overallProgress
        ]
        
        if let currentFile = progress.currentFile {
            eventData["currentFile"] = currentFile
            eventData["fileProgress"] = progress.fileProgress
        }
        
        fireEvent("progress", with: eventData)
    }
    
    private func handleFileProgress(filename: String, sent: Int64, total: Int64) {
        guard isActive else { return }
        
        let progress = Double(sent) / Double(total)
        
        fireEvent("fileprogress", with: [
            "filename": filename,
            "sent": sent,
            "total": total,
            "progress": progress
        ])
    }
    
    private func handleComplete(statusCode: Int, headers: [String: String], data: Data?) {
        guard isActive else { return }
        
        isActive = false
        
        var eventData: [String: Any] = [
            "statusCode": statusCode,
            "headers": headers,
            "success": statusCode >= 200 && statusCode < 300
        ]
        
        if let data = data, let response = String(data: data, encoding: .utf8) {
            eventData["body"] = response
        }
        
        TNMLogger.success("Multipart upload completed", feature: "MultipartUpload", details: [
            "uploadId": uploadId,
            "statusCode": statusCode
        ])
        
        fireEvent("complete", with: eventData)
    }
    
    private func handleError(_ error: Error) {
        guard isActive else { return }
        
        isActive = false
        
        TNMLogger.error("Multipart upload error", feature: "MultipartUpload", error: error, details: [
            "uploadId": uploadId
        ])
        
        fireEvent("error", with: [
            "error": error.localizedDescription,
            "code": (error as NSError).code
        ])
    }
}

File: android/src/ti/network/manager/TNMMultipartUploadProxy.java
/**
 * Ti.Network.Manager - Multipart Upload Proxy
 * JavaScript proxy for multipart uploads
 * Mirrors iOS TNMMultipartUploadProxy
 */

package ti.network.manager;

import org.appcelerator.kroll.KrollDict;
import org.appcelerator.kroll.KrollProxy;
import org.appcelerator.kroll.annotations.Kroll;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@Kroll.proxy(creatableInModule = TiNetworkManagerModule.class)
public class TNMMultipartUploadProxy extends KrollProxy {

    private final TNMMultipartUploadManager uploadManager;
    private final String uploadId;
    private final String url;
    private final Map<String, String> headers;
    private final List<TNMMultipartUploadManager.MultipartField> fields;
    private final String priority;
    private boolean isActive = false;

    public TNMMultipartUploadProxy(KrollDict params, TNMMultipartUploadManager uploadManager) {
        super();

        this.uploadManager = uploadManager;
        this.uploadId = UUID.randomUUID().toString();
        this.url = params.optString("url", "");
        this.priority = params.optString("priority", "normal");

        // Headers
        if (params.containsKey("headers")) {
            @SuppressWarnings("unchecked")
            Map<String, String> headersMap = (Map<String, String>) params.get("headers");
            this.headers = headersMap;
        } else {
            this.headers = new HashMap<>();
        }

        // Parse fields
        this.fields = new ArrayList<>();
        if (params.containsKey("fields")) {
            Object[] fieldsArray = (Object[]) params.get("fields");

            assert fieldsArray != null;
            for (Object fieldObj : fieldsArray) {
                @SuppressWarnings("unchecked")
                Map<String, Object> field = (Map<String, Object>) fieldObj;

                String type = (String) field.get("type");
                String name = (String) field.get("name");

                if ("text".equals(type)) {
                    // Text field
                    String value = (String) field.get("value");
                    fields.add(new TNMMultipartUploadManager.MultipartField(name, value));

                } else if ("file".equals(type)) {
                    // File field
                    String filename = (String) field.get("filename");
                    String mimeType = (String) field.get("mimeType");
                    String data = (String) field.get("data");

                    fields.add(new TNMMultipartUploadManager.MultipartField(
                            name, filename, mimeType, data
                    ));
                }
            }
        }

        Map<String, Object> details = new HashMap<>();
        details.put("uploadId", uploadId);
        details.put("url", url);
        details.put("fieldCount", fields.size());
        TNMLogger.debug("Multipart upload proxy created", "MultipartUpload", details);
    }

    @Kroll.method
    public void upload() {
        if (isActive) {
            TNMLogger.warning("Upload already active", "MultipartUpload",
                    Map.of("uploadId", uploadId));
            return;
        }

        isActive = true;

        // Convert priority
        float urlPriority = 0.5f; // Normal
        if ("high".equals(priority)) {
            urlPriority = 0.75f;
        } else if ("low".equals(priority)) {
            urlPriority = 0.25f;
        }

        uploadManager.upload(
                uploadId,
                url,
                headers,
                fields,
                urlPriority,
                new TNMMultipartUploadManager.UploadCallback() {
                    @Override
                    public void onProgress(long sent, long total, double progress, String currentFile) {
                        handleProgress(sent, total, progress, currentFile);
                    }

                    @Override
                    public void onFileProgress(String filename, double progress) {
                        handleFileProgress(filename, progress);
                    }

                    @Override
                    public void onComplete(int statusCode, Map<String, String> headers, String body) {
                        handleComplete(statusCode, headers, body);
                    }

                    @Override
                    public void onError(Exception error) {
                        handleError(error);
                    }
                }
        );
    }

    @Kroll.method
    public void cancel() {
        if (!isActive) return;

        uploadManager.cancel(uploadId);
        isActive = false;

        KrollDict event = new KrollDict();
        fireEvent("cancelled", event);
    }

    private void handleProgress(long sent, long total, double progress, String currentFile) {
        if (!isActive) return;

        KrollDict event = new KrollDict();
        event.put("uploadId", uploadId);
        event.put("totalBytesSent", sent);
        event.put("totalBytesExpected", total);
        event.put("progress", progress);
        if (currentFile != null) {
            event.put("currentFile", currentFile);
        }

        fireEvent("progress", event);
    }

    private void handleFileProgress(String filename, double progress) {
        if (!isActive) return;

        KrollDict event = new KrollDict();
        event.put("uploadId", uploadId);
        event.put("filename", filename);
        event.put("progress", progress);

        fireEvent("fileprogress", event);
    }

    private void handleComplete(int statusCode, Map<String, String> headers, String body) {
        if (!isActive) return;

        isActive = false;

        KrollDict event = new KrollDict();
        event.put("uploadId", uploadId);
        event.put("statusCode", statusCode);
        event.put("headers", headers);
        event.put("body", body != null ? body : "");
        event.put("success", statusCode >= 200 && statusCode < 300);

        fireEvent("complete", event);
    }

    private void handleError(Exception error) {
        if (!isActive) return;

        isActive = false;

        KrollDict event = new KrollDict();
        event.put("uploadId", uploadId);
        event.put("error", error.getMessage());

        fireEvent("error", event);
    }
}